aiida_quantumespresso/parsers/pp.py

```diff
@@ -0,0 +1,332 @@
+        if 'ERROR_OUTPUT_STDOUT_INCOMPLETE' in logs['error']:
+            return self.exit_codes.ERROR_OUTPUT_STDOUT_INCOMPLETE
+
+        # Parse the post-processed-data according to what kind of data file was produced
+        if output_parameters['output_format'] == 'gnuplot':
+            parsed_data = self.parse_gnuplot1D(data_raw)
+        elif output_parameters['output_format'] == 'gnuplot x,y,f':
+            parsed_data = self.parse_gnuplot2D(data_raw)
+        elif output_parameters['output_format'] == 'Gaussian cube':
+            parsed_data = self.parse_gaussian(data_raw)
+        else:
+            return self.exit_codes.ERROR_UNSUPPORTED_DATAFILE_FORMAT
+
+        # Create output nodes
+        self.out('output_data', parsed_data)
+        self.out('output_parameters', orm.Dict(dict=output_parameters))
+
+
+    def parse_stdout(self, stdout_str):
+        """
+        Parses the output written to StdOut to retrieve basic information about the post processing
+
+        :param stdout_str: the stdout file read in as a single string
+        """
+
+        def detect_important_message(logs, line):
+            """
+            Detect know errors and warnings printed in the stdout
+
+            :param logs:
+            :param line: a line from the stdout as a string
+            """
+            message_map = {
+                'error': {
+                    'xml data file not found': 'ERROR_PARENT_XML_MISSING'
+                },
+                'warning': {
+                    'Warning:': None,
+                    'DEPRECATED:': None,
+                }
+            }
+
+            # Match any known error and warning messages
+            for marker, message in message_map['error'].items():
+                if marker in line:
+                    if message is None:
+                        message = line
+                    logs.error.append(message)
+
+            for marker, message in message_map['warning'].items():
+                if marker in line:
+                    if message is None:
+                        message = line
+                    logs.warning.append(message)
+
+
+        stdout_lines = stdout_str.splitlines()
+        logs = get_logging_container()
+        output_dict = {}
+
+        # Check for job completion, indicating that pp.x exited without interruption, even if there was an error.
+        for line in stdout_lines:
+            if 'JOB DONE' in line:
+                break
+        else:
+            logs.error.append('ERROR_OUTPUT_STDOUT_INCOMPLETE')
+
+        # Detect any issues and detect job completion
+        for line in stdout_lines:
+            detect_important_message(logs, line)
+
+        # Parse useful data from stdout
+        for line in stdout_lines:
+            if 'Check:' in line:
+		if 'negative/imaginary' in line:    # QE6.1
+                    output_dict['negative_core_charge'] = float(line.split('=')[-1].split()[0])
```

Comment:
small optimizations in these lines is too call `line.split()` once and then use that instead of calling multiple times

Reply:
Done.